src/Lean/Linter/UnusedVariables.lean

```diff
@@ -1,34 +1,109 @@
+metavariable context looking for occurrences of `x`. We use caching to ensure that this is still
+linear in the size of the info tree, even though there are many metavariable contexts in all the
+intermediate stages of elaboration; these are highly similar and make use of `PersistentHashMap`
+so there is a lot of subterm sharing we can take advantage of.
+
+## The `@[unused_variables_ignore_fn]` attribute
+
+Some occurrences of variables are deliberately unused, or at least we don't want to lint on unused
+variables in these positions. For example:
+
+```
+def foo (x : Nat) : (y : Nat) → Nat := fun _ => x
+                  -- ^ don't lint this unused variable because it is public API
+```
+
+They are generally a syntactic criterion, so we allow adding custom `IgnoreFunction`s so that
+external syntax can also opt in to lint suppression, like so:
+
+```
+macro (name := foobarKind) "foobar " name:ident : command => `(def foo ($name : Nat) := 0)
+
+foobar n -- linted because n is unused in the macro expansion
+
+@[unused_variables_ignore_fn]
+def ignoreFoobar : Lean.Linter.IgnoreFunction := fun _ stack _ => stack.matches [``foobarKind]
+
+foobar n -- not linted
+```
+
+-/
 
 namespace Lean.Linter
 open Lean.Elab.Command Lean.Server Std
 
+/-- Enables or disables all unused variable linter warnings -/
 register_builtin_option linter.unusedVariables : Bool := {
   defValue := true,
   descr := "enable the 'unused variables' linter"
 }
+/-- Enables or disables unused variable linter warnings in function arguments -/
 register_builtin_option linter.unusedVariables.funArgs : Bool := {
   defValue := true,
   descr := "enable the 'unused variables' linter to mark unused function arguments"
 }
+/-- Enables or disables unused variable linter warnings in patterns -/
 register_builtin_option linter.unusedVariables.patternVars : Bool := {
   defValue := true,
   descr := "enable the 'unused variables' linter to mark unused pattern variables"
 }
 
-def getLinterUnusedVariables (o : Options) : Bool := getLinterValue linter.unusedVariables o
-def getLinterUnusedVariablesFunArgs (o : Options) : Bool := o.get linter.unusedVariables.funArgs.name (getLinterUnusedVariables o)
-def getLinterUnusedVariablesPatternVars (o : Options) : Bool := o.get linter.unusedVariables.patternVars.name (getLinterUnusedVariables o)
+/-- Gets the status of `linter.unusedVariables` -/
+def getLinterUnusedVariables (o : Options) : Bool :=
+  getLinterValue linter.unusedVariables o
+
+/-- Gets the status of `linter.unusedVariables.funArgs` -/
+def getLinterUnusedVariablesFunArgs (o : Options) : Bool :=
+  o.get linter.unusedVariables.funArgs.name (getLinterUnusedVariables o)
 
+/-- Gets the status of `linter.unusedVariables.patternVars` -/
+def getLinterUnusedVariablesPatternVars (o : Options) : Bool :=
+  o.get linter.unusedVariables.patternVars.name (getLinterUnusedVariables o)
+
+/-- An `IgnoreFunction` receives:
+
+* a `Syntax.ident` for the unused variable
```

Comment:
Time to make this an `Ident`? Potential breaking change but downstream users most likely just do a lambda without type annotations; in fact, all uses on GitHub don't use the ident at all.

Reply:
The only one that used the ident to begin with was the check for `_x` variables, but this has been incorporated into the main algorithm.

Reply:
Based on the way in which this syntax was produced, I'm not sure we can actually say it is definitely an `Ident`. It is the declStx with declId stripped, but if it happened to be some other syntax like an atom that an elaborator decided to use as the span of a variable I think we could end up here with a non- `.ident`. We pass a syntax stack corresponding to an `.ident` but I think the `declStx` doesn't necessarily have to be the same thing as the head of the stack, they are only matched up using syntax ranges.

Reply:
Ok, let's leave it as is then